### ISSUE TITLE
hx freezes on file picker selection

### ISSUE THREAD

**Issue by ondrej-ivanko** [helix-editor/helix#5871]
### Summary

Execturing `hx` from certain folders seems to freeze file picker (SPC f). I tried to run `hx` from directory one level lower than /home dir (/home/Downloads) and when trying to run file picker  with _space_ + _f_ it froze `hx` and I couldn' t do any action besides shutting down terminal.

This issue is resolved by putting the folder that's in hx logs (/Games) to .gitignore. My theory is that path is just too long for `hx` to handle (Input/Output Err in logs). Or that there is problem with permissions, as some files in my /home dir are owned by root. Not sure why they are owned by root, but they're managed by wine application, so I'm not going to change the owner to not mess up with wine.

### Reproduction Steps

I tried this:

1. `hx`
2. press space and f

I expected this to happen:
file picker pops up with no issue

Instead, this happened:
hx froze and stops responding

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-02-07T12:47:56.800 helix_term::commands [ERROR] Global search error: /home/oji/Games/origin/dosdevices/z:/sys/module/snd_pcm_dmaengine/holders/snd_soc_core/drivers/platform:snd-soc-dummy/snd-soc-dummy/subsystem/drivers/tigerlake-pinctrl/INTC1055:00/firmware_node/subsystem/devices/ACPI0007:0e/physical_node/subsystem/drivers/processor/cpu0/node0/cpu13/firmware_node/thermal_cooling/subsystem/thermal_zone11/hwmon4/subsystem/hwmon1/device/subsystem/ADP0/wakeup29/subsystem/wakeup8/device/iommu/subsystem/dmar0/devices/0000:00:02.0/drm/controlD64/card0-DP-1/i2c-12/subsystem/drivers/i2c_hid_acpi/i2c-MSFT0001:01/0018:06CB:CE78.0005/input/input11/subsystem/event2/device/device/input/input2/power/autosuspend_delay_ms, Input/output error (os error 5)
```
```
2023-02-07T12:47:56.735 helix_term::commands [ERROR] Global search error: /home/oji/Games/origin/dosdevices/z:/sys/module/snd_pcm_dmaengine/holders/snd_soc_core/drivers/platform:snd-soc-dummy/snd-soc-dummy/subsystem/drivers/tigerlake-pinctrl/INTC1055:00/firmware_node/subsystem/devices/ACPI0007:0e/physical_node/subsystem/drivers/processor/cpu0/node0/cpu13/firmware_node/thermal_cooling/subsystem/thermal_zone11/hwmon4/subsystem/hwmon1/device/subsystem/ADP0/wakeup29/subsystem/wakeup8/device/iommu/subsystem/dmar0/devices/0000:00:02.0/drm/controlD64/card0-DP-1/i2c-12/subsystem/drivers/i2c_hid_acpi/i2c-MSFT0001:01/0018:06CB:CE78.0005/input/input11/subsystem/input18/device/device:18/physical_node/subsystem/card1/device/i2c-19/delete_device, Permission denied (os error 13)
```

</details>

### Platform

Linux - Fedora 37 86_64

### Terminal Emulator

Alacritty 0.10.1

### Helix Version

helix 22.12-223-g9c98043c (9c98043c)

**Comment by pascalkuthe** [comment]
The log entry you attached is related to global search `<space>g`. Not the file picker `<space>f`.  However it sounds like you got an entire filesystem under that directory. There could simply be a lot of files there which take a long time to transverse. 

However it's a bit odd that this occurs here since we limit the number of files outside of git repos to 100k currently. Perhaps that specific subfilesystem is somekind of fuse filesystem that is responding very slowly.

**Comment by ondrej-ivanko** [comment]
Hi pascal,

thank you for the swift response. In that case, I can only assume `hx` freezes and does not write any useful information to logs, if those logs are related to global search (which in my case is assigned to `<space>/`). I also tried to run it with `RUST_BACKTRACE=full hx`. Makes no difference. Nothing in logs.

I remember global search also froze at least once, but I cant reproduce that situation again.

**Comment by CptPotato** [comment]
I can somewhat reproduce this if I try `<space>f` or `<space>/` on a network drive (with a ton of subfolders). I assume walking the files is simply too slow in this case.

If walking the files takes more than a few seconds, it might be neat to have some visual feedback and be able to `C-c` to cancel:
![image](https://user-images.githubusercontent.com/3957610/217521756-554a8310-4bdb-437b-a278-e85990d2deff.png)

Though there might be another issue besides slow file access that @ondrej-ivanko is experiencing.
I tried to see if access rights could cause this, but the file picker handled this just fine (tested on Windows).

**Comment by pascalkuthe** [comment]
> I can somewhat reproduce this if I try `<space>f` or `<space>/` on a network drive (with a ton of subfolders). I assume walking the files is simply too slow in this case.
> 
> If walking the files takes more than a few seconds, it might be neat to have some visual feedback and be able to `C-c` to cancel: ![image](https://user-images.githubusercontent.com/3957610/217521756-554a8310-4bdb-437b-a278-e85990d2deff.png)
> 
> Though there might be another issue besides slow file access that @ondrej-ivanko is experiencing. I tried to see if access rights could cause this, but the file picker handled this just fine (tested on Windows).

Yeah that would be nice. Some prior discussion/work #1987, #1543 and https://github.com/helix-editor/helix/issues/1707.

I am also working on a watchmen like filewatcher in rust we could use for helix which would keep a cached a copy of the workspace filetree in memory (and update it automatically either based on native os events or polling as a fallback). So that would mean that this slow task is atleast only performed once and in the background.

That being said even such a background indexer could stall and therefore not return anything for a longtime. There will also be options/heuristics to disable this indexer. In both cases an iteractive hint like you showed there would be very nice.

We could probably compute pickers in a `job` instead.

**Comment by pascalkuthe** [comment]
> 

To allow the option to respond to keyevents (`C-c`)  or even just show this stuatus you must run the task in the background as only the main loop is the only thing that can render or respond to key events. It's also important to keep the mainloop running so we can respond to other async events like signals or resizing the window.

Freezing the editor while waiting on certain jobs would be more of a followup feature in my opinion and could also be used for other async tasks like `goto_reference` to make it less confusing. So this would be thtree seperate features:
* Make the file picker/global search async
* Optimize the file picker/global search by using a filewatcher and an in memory fs tree
* softlock the editor while waiting for async requests (like filepicker, globalsearch and `goto_reference`) that are needed to fullfill user actions

**Comment by ondrej-ivanko** [comment]
I agree with the proposition to allow the user stop the `file picker` process with some key-event. Another option would be some sort of short timeout to load files to file picker, but user should be informed about this timeout being in effect. If it would be really short, than I guess user wouldn't care waiting like 2s.

The solution should most importantly not force the user to shut down the whole terminal. I'm using zellij multiplexer and running `hx` in one of two tabs. The terminal froze and I couldn't even switch to different tab and had to force shut terminal or kill hx process from gnome-terminal.

**Comment by kristian-clausal** [comment]
Global search (space /) also freezes when I forget to specify the right working directory (I assume it would finish at some point), it would be really nice to not have to kill the terminal or the pane to cancel it.

**Comment by catdevnull** [comment]
i believe this happened for me because i had a `.git` in my projects folder which caused it to try and search through all of my projects. deleting it fixes it

**Comment by mway** [comment]
There are times, though, where it's unavoidable to have a large tree (e.g. monorepos) - maybe it's worth adding an option to only lazily search based on input, instead of pre-traversing everything when the modal opens?

In my case, if I type quickly enough (so as to immediately begin filtering), the picker is responsive; if I just cmd+f and type nothing, it will take many seconds (sometimes up to 20) to finish loading, because it traverses the entire monorepo. This isn't a huge deal because you can cmd+F as well, except when I typo and think that I've hit cmd+F but have actually hit cmd+f. (Personally, I think the two shortcuts should be reversed: cmd+f for rooting at pwd, cmd+F for rooting at the project root).

I haven't had a chance to dig yet, but assuming there's more going on behind the scenes than just stat'ing files (e.g. if they're eagerly being loaded into memory for previews), maybe an option to disable some of that work would be enough?

**Comment by webbb82** [comment]
I have this same error/freeze. As far as i can tell it only happens when trying to find when you are in the space menu the space f and space F command the second f/F is pressed helix will lock up. only way to get out of it is to force close the terminal. This started happening recently maybe a update ago. I use helix everyday so Im pretty sure its only been a week or so of this issue. 
Im not sure if this is related but i also started seeing these ilttle red lines on the left hand side. seem to be connected to the re circles that have to do with maybe compositor ? i cant be sure. prob not related. 
Im running arch linux
- with wayland Hyprland
- Nvidia graphics card
- intel chick
- zen kernel 
hope this helps as i would like to see this resolved.
